tests/gtest/avifrgbtoyuvtest.cc

```diff
@@ -0,0 +1,387 @@
+template <typename PixelType>
+void ModifyImageChannel(avifRGBImage* image, uint32_t channel_offset,
+                        const int32_t modifier[]) {
+  const uint32_t channel_count = avifRGBFormatChannelCount(image->format);
+  assert(channel_offset < channel_count);
+  for (uint32_t y = 0, i = 0; y < image->height; ++y) {
+    PixelType* pixel =
+        reinterpret_cast<PixelType*>(image->pixels + image->rowBytes * y);
+    for (uint32_t x = 0; x < image->width; ++x, ++i) {
+      pixel[channel_offset] += modifier[i];
+      pixel += channel_count;
+    }
+  }
+}
+
+void ModifyImageChannel(avifRGBImage* image, uint32_t channel_offset,
+                        const int32_t modifier[]) {
+  (image->depth <= 8)
+      ? ModifyImageChannel<uint8_t>(image, channel_offset, modifier)
+      : ModifyImageChannel<uint16_t>(image, channel_offset, modifier);
+}
+
+// Fills the image channel with the given value, and modifies the individual
+// pixel values of that channel with the modifier, if not null.
+void SetImageChannel(avifRGBImage* image, uint32_t channel_offset,
+                     uint32_t value, const int32_t modifier[]) {
+  testutil::FillImageChannel(image, channel_offset, value);
+  if (modifier) {
+    ModifyImageChannel(image, channel_offset, modifier);
+  }
+}
+
+// Accumulates stats about the differences between the images a and b.
+template <typename PixelType>
+void GetDiffSumAndSqDiffSum(const avifRGBImage& a, const avifRGBImage& b,
+                            int64_t* diff_sum, int64_t* abs_diff_sum,
+                            int64_t* sq_diff_sum, int64_t* max_abs_diff) {
+  const uint32_t channel_count = avifRGBFormatChannelCount(a.format);
+  for (uint32_t y = 0; y < a.height; ++y) {
+    const PixelType* row_a =
+        reinterpret_cast<PixelType*>(a.pixels + a.rowBytes * y);
+    const PixelType* row_b =
+        reinterpret_cast<PixelType*>(b.pixels + b.rowBytes * y);
+    for (uint32_t x = 0; x < a.width * channel_count; ++x) {
+      const int64_t diff =
+          static_cast<int64_t>(row_b[x]) - static_cast<int64_t>(row_a[x]);
+      *diff_sum += diff;
+      *abs_diff_sum += std::abs(diff);
+      *sq_diff_sum += diff * diff;
+      *max_abs_diff = std::max(*max_abs_diff, std::abs(diff));
+    }
+  }
+}
+
+void GetDiffSumAndSqDiffSum(const avifRGBImage& a, const avifRGBImage& b,
+                            int64_t* diff_sum, int64_t* abs_diff_sum,
+                            int64_t* sq_diff_sum, int64_t* max_abs_diff) {
+  (a.depth <= 8) ? GetDiffSumAndSqDiffSum<uint8_t>(a, b, diff_sum, abs_diff_sum,
+                                                   sq_diff_sum, max_abs_diff)
+                 : GetDiffSumAndSqDiffSum<uint16_t>(
+                       a, b, diff_sum, abs_diff_sum, sq_diff_sum, max_abs_diff);
+}
+
+// Returns the Peak Signal-to-Noise Ratio from accumulated stats.
+double GetPsnr(double sq_diff_sum, double num_diffs, double max_abs_diff) {
+  if (sq_diff_sum == 0.) {
```

Comment:
[Just a comment. No change requested.] I seem to remember there was a style cleanup effort in Chrome or Google's internal repository to change all floating-point constants of the form `ddd.` to `ddd.0`. Would be good to look into that.